### ISSUE TITLE
How to save model parameters in different ranks? (especially for MOE model)

### ISSUE THREAD

**Issue by Luodian** [facebookresearch/fairscale#1017]
I’ve implemented MoE layer on my codebase. I used 32 experts in total on 8 GPUs, but when I check the model, it only displays 32 // 8 = 4 experts. But the code works fine and gate’s auxillary loss went down well. My implementation is in: https://gist.github.com/Luodian/055993c44f2a80a2a7440cf2ad262ac8

However, I found if I directly use “torch.save_model()”, I can just save 4 experts to checkpoint file. So I guess that I should save parameters on different ranks myself. But I didnt have an idea how to implement it and saw you mentioned something related on github.

Is my guess right? And I am wondering if fairscale has provided anything explaining how to implement this or where I can find some documents or tutorials to realize this operation.

**Comment by min-xu-ai** [comment]
There isn't very good document on this yet. In general, you can call module.local_state_dict on each rank if you are using FSDP. `load_local_state_dict` can be used to load back the checkpoint files. Pytorch now has a newer version of FSDP as well, please check with pytorch folks as well.

**Comment by Luodian** [comment]
Thanks Min! 

I've tried this way and found `model.local_state_dict() ` can save flatten params in different ranks and supposedly they are right ones. But how could I load them back to form a complete MoE model? Is it just simply call load_local_state_dict or should I do more processing operations?
![WX20220624-121646](https://user-images.githubusercontent.com/15847405/175449267-9330be88-fbb0-4cb4-83dc-73dbadcd1751.png)

**Comment by min-xu-ai** [comment]
I think loading them directly should work as long as the same number of GPUs is used. Pls let me know if you run into issues.

**Comment by Luodian** [comment]
Thanks Min~

I implemented the way using loading `local_state_dict()`, and I see no errors when running, and losses continue decrease at the point when they are saved. However, I have question: in FSDP (but without OSS), should I also save & load optimizers in different ranks? 

my save & load codes are:
```python
def save_fsdp_model(args, epoch, model, model_without_ddp, optimizer, loss_scaler, msg=None):
    output_dir = Path(args.output_dir)
    checkpoint_dir_path = output_dir / f'checkpoint-{msg}-epoch-{epoch}'
    Path(checkpoint_dir_path).mkdir(parents=True, exist_ok=True)
    checkpoint_path = checkpoint_dir_path / f'rank-{get_rank()}.pth'
    ### save model on different ranks to different files ###
    to_save = {'model': model.local_state_dict(), 'optimizer': optimizer.state_dict(), 'epoch': epoch, 'scaler': loss_scaler.state_dict(), 'args': args, }
    torch.save(to_save, checkpoint_path)

def load_fsdp_model(args, model, optimizer, loss_scaler):
    if args.resume:
        checkpoint_dir = Path(args.resume)
        checkpoint_path = checkpoint_dir / f'rank-{get_rank()}.pth'
        checkpoint = torch.load(checkpoint_path)
        model.load_local_state_dict(checkpoint['model'])
        print("Resume checkpoint %s" % checkpoint_path)
        ### load model in different ranks with corresponding checkpoints ###
        if 'optimizer' in checkpoint and 'epoch' in checkpoint and not (hasattr(args, 'eval') and args.eval):
            optimizer.load_state_dict(checkpoint['optimizer'])
            args.start_epoch = checkpoint['epoch'] + 1
            if 'scaler' in checkpoint:
                loss_scaler.load_state_dict(checkpoint['scaler'])
            print("With optim & sched!")
```

my optimizer related codes are:

```python
    param_groups = optim_factory.add_weight_decay(model, args.weight_decay)
    optimizer = torch.optim.AdamW(param_groups, lr=args.lr, betas=(0.9, 0.95))

**Comment by min-xu-ai** [comment]
I do think you need save/load optimizer state in different ranks. FSDP splits parameters across ranks, hence each rank have different optimizer state for different params. In a way, FSDP has a built-in OSS.

**Comment by Luodian** [comment]
Thanks Min!

However, I tried to find a way to save optimizer states on different ranks. But I can't find a direct way like a `optim_state = optimizer.local_state_dict()` and just save it? Is there any related documents or codes provided in fairscale? Or if the optimizer has been sharded, then `optim_state = optimizer.state_dict()` is the local state on corresponding rank?

**Comment by min-xu-ai** [comment]
Yes, it is already sharded. So directly call state_dict() should be fine.

**Comment by min-xu-ai** [comment]
Please reopen if still issues.
